[Angular - NgRx - SCSS] Single Repo API Setup

lindakatcodes opened the pull request:
Closes issue #183 

There's a few things still to decide on and solve for this.
- Ran into some issues with testing for the username (it ends up being a nested property in the call), but I think even if we don't pull that info from this call, we'll already have it for the user. Just needs to actually be set when the UI is set up. I think that field can either be deleted from the AC or else I'll add it in when I come back.
- The other unmarked fields will all require extra fetch calls, since we're using the REST API. Decisions on which we want and which we don't will need to be made. This PR isn't fully ready until those choices are made.  

Acceptance criteria from ticket:

data needed:
- [x]  username
- [x]  repo name
- [x]  visibility badge (public / private)
- [x]  watch count
- [x]  star count
- [x]  fork count
- [x]  issue count
- [x]  pull request count
- [ ]  branch names
- [x]  about description
- [x]  about website link (if available)
- [x]  root tree information (folders and files)
- [x]  readme information

lindakatcodes commented:
6/20 update:

- This still needs to have the tests updated and completed, but the main functionality is now complete. 
- Branch names were decided not to be included at this time. None of the other showcases actively use them, and it would require a separate fetch call for information that's not actively used at the moment.

ihardz commented:
replacing type **any** with **unknown** can help to avoid the Github Linting Action warnings...

Reshurum commented:
I noticed branch names are not included. Is this impossible or impractical to support with the REST API?

Reshurum commented on angular-ngrx-scss/src/app/repository/services/repository.service.ts:
```diff
@@ -0,0 +1,79 @@
+    );
+  }
+
+  // TODO: readme file is currently an encoded string - this method should be improved to either return the raw data, or include the fields needed to de-code the string when ready to display it
+  // TODO: write test for this function when it's updated
+  getReadmeContent(owner: string, repoName: string): Observable<string> {
+    const url = `${environment.githubUrl}/repos/${owner}/${repoName}/readme`;
+
+    return this.http
+      .get<ReadmeApiResponse>(url)
+      .pipe(map((file) => file.content));
+  }
```
Do we know if any of the other starters are using a specific README rendering library? I am curious if a decision was already made at some point as to which one we should use.

lindakatcodes replied:
Not sure. Since this ticket is just setting up the API and not worrying about rendering the document I left it like this, but would be worth looking into and adjusting when it comes time to actually show the README! There has to be something, I'd think.

lindakatcodes commented on angular-ngrx-scss/.eslintrc.json:
```diff
@@ -5,7 +5,7 @@
       "files": ["*.ts"],
       "parserOptions": {
-        "project": ["tsconfig.json"],
+        "project": ["**/tsconfig.json"],
```
I consistently see ESLint errors for the top level imports in nested component files. Adjusting this fixed those. I think the nested component files especially have trouble reaching up far enough to get the tsconfig? Moreso than the tsconfig itself being nested.

Reshurum replied:
I cannot reproduce this locally when I remove the wildcard. Does this still happen for you with a fresh clone or after cleaning your repo of untracked files? Running `npm run lint` without this change only reveals 5 warnings about "any" and unused variables for me.

lindakatcodes replied:
I can revert this. I can't find the specific instances anymore and it's not truly an issue for this PR. Possible my editor was just being wonky.

Reshurum commented:
> Adjusted this in one location where it works well, but it creates more errors than it solves for all the jasmine spy mocks. In the jasmine types, they have all the spy objects set to return `any` values, so I think we're just sort of stuck with using `any` for those. I added an es-lint disable line in those instances to keep things clearer. Since it's only affecting tests, I think this is a reasonable use case.

The issue is the automatic type resolution for `jasmine.createSpyObj` results in the variant that returns `any`. The typings for this function indicate it can also return a SpyObj<T>. The use of `any` here can be replaced with `SpyObj<T>` where `T` is the Angular service you are mocking.

lindakatcodes commented:
> > Adjusted this in one location where it works well, but it creates more errors than it solves for all the jasmine spy mocks. In the jasmine types, they have all the spy objects set to return `any` values, so I think we're just sort of stuck with using `any` for those. I added an es-lint disable line in those instances to keep things clearer. Since it's only affecting tests, I think this is a reasonable use case.
> 
> The issue is the automatic type resolution for `jasmine.createSpyObj` results in the overload that returns `any`. The typings for this function indicate it can also return a SpyObj. The use of `any` here can be replaced with `SpyObj<T>` where `T` is the Angular service you are mocking.

Oh, I like that better, thanks Jamie! Adjusted for all but 1 - there's something off with the naming of things in the auth spec, and since I'll have another PR soon dealing with that area of the codebase, I'll adjust that one there.
